ra2ce/network/network_simplification/snkit_network_merge_wrapper.py

```diff
@@ -94,62 +98,159 @@ def _get_merged_edges(
+    def filter_node(_node_set: set, _degrees: int) -> set:
+        _filtered = set()
+        for _node_id in _node_set:
+            # Get the predecessors (antecedents) and successors (precedents) to make sure for to filter correctly
+            # For _node_set with all degree 2: this filters on the nodes that have only one predecessor and successor.
+            # E.g. filters on 2 in 1->2, 1->5, 2->3, 3->4, 3->5
+            # For _node_set with all degree 4: Check if the predecessors and successors are the same nodes.
+            # E.g. filters on 2 in 1->2, 2->3, 2->1, 3->2.
+            predecessors = list(networkx_graph.predecessors(_node_id))
+            successors = list(networkx_graph.successors(_node_id))
+
+            # Check the degree of the _node_set and the corresponding criterium.
+            if ((_degrees == 2 and len(predecessors) == len(successors) == 1) or
+                    (_degrees == 4 and sorted(predecessors) == sorted(successors))):
+                _filtered.add(_node_id)
+        return _filtered
+
+    def get_edge_paths(node_set: set, _snkit_network: SnkitNetwork) -> list:
+        def get_adjacency_list(edges_gdf: gpd.GeoDataFrame, from_id_column: str = "from_id",
+                                to_id_column: str = "to_id") -> defaultdict:
+            # Convert the edges of a GeoDataFrame to an adjacency list using vectorized operations.
+            _edge_dict = defaultdict(set)
+            # Extract the 'from_id' and 'to_id' columns as numpy arrays for efficient processing
+            from_ids = edges_gdf[from_id_column].values
+            to_ids = edges_gdf[to_id_column].values
+            # Vectorized operation to populate the adjacency list
+            for from_id, to_id in np.nditer([from_ids, to_ids]):
+                _edge_dict[int(from_id)].add(int(to_id))
+                _edge_dict[int(to_id)].add(int(from_id))
+
+            return _edge_dict
+
+        def retrieve_edge(node1: int | float, node2: int | float) -> gpd.GeoDataFrame:
+            """Retrieve the edge from snkit_network.edges GeoDataFrame between two nodes."""
+            edge = _snkit_network.edges[
+                (_snkit_network.edges['from_id'] == node1) &
+                (_snkit_network.edges['to_id'] == node2)
+                ]
+            return edge if not edge.empty else None
+
+        def construct_path(start_node: int | float, end_node: int | float, __intermediates: list) -> pd.DataFrame | None:
+            path = []
+            current_node = start_node
+            _intermediates = __intermediates.copy()
+
+            while _intermediates:
+                for next_node in _intermediates:
+                    edge = retrieve_edge(current_node, next_node)
+                    if edge is not None:
+                        path.append(edge)
+                        _intermediates.remove(next_node)
+                        current_node = next_node
+                        break
```

Comment:
`while` loops are very dangerous as they are prone to infinite loops. I do not fully understand the purpose of this algorithm.
Try the following which also removes the need of making a copy (which is very expensive for memory and computation). Keep in mind I did not want to invest much time in rethinking the algorithm, could be improved, but it should meet your requirements.

```suggestion
            _explored_nodes = []
            # Ensure we go through all the items.
            for _ in range(len(intermediate_nodes)):
                # Filter out nodes already used for edge retrieval
                _to_explore = filter(lambda x: x not in _explored_nodes, intermediate_nodes)
                for _next_node in _to_explore:
                    _edge = retrieve_edge(current_node, _next_node)
                    if _edge is not None:
                        path.append(_edge)
                        _explored_nodes.append(_next_node)
                        current_node = _next_node
```

Reply:
modified